### PR TITLE
Refactor of Cache.py

### PR DESCRIPTION
**Pull request by matusvalo**
Fixes of comments from #6090

### CODE REVIEW

**Comment by matusvalo** [comment]
Not all comments were implemented therefor PR is a draft.

**Comment by matusvalo** [comment]
99% comments were fixed in #6090 so marking it as ready for review...

File: Cython/Build/Dependencies.py [hunk 1]
```diff
@@ -1237,7 +1251,7 @@ def cythonize_one(pyx_file, c_file, fingerprint, cache, quiet, options=None,
             raise CompileError(None, pyx_file)
         elif os.path.exists(c_file):
             os.remove(c_file)
-    elif fingerprint:
+    elif cache and fingerprint:
         cache.store_to_cache(fingerprint, c_file, result)
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply Cython/Build/Dependencies.py]: fixed.

File: Cython/Build/Cache.py [hunk 2]
```diff
@@ -118,9 +132,8 @@ def fingerprint_file_base(self, cfile, fingerprint):
     def gz_fingerprint_file(self, cfile, fingerprint):
         return self.fingerprint_file_base(cfile, fingerprint) + gzip_ext
 
-
     def zip_fingerprint_file(self, cfile, fingerprint):
-        return self.fingerprint_file_base(cfile, fingerprint) + '.zip'
+        return self.fingerprint_file_base(cfile, fingerprint) + zip_ext
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply Cython/Build/Cache.py]: fixed.

File: Cython/Build/Cache.py [hunk 3]
```diff
@@ -131,44 +144,58 @@ def lookup_cache(self, c_file, fingerprint):
...
-                os.utime(zip_fingerprint_file, None)
-                dirname = os.path.dirname(c_file)
-                with zipfile.ZipFile(zip_fingerprint_file) as z:
-                    for artifact in z.namelist():
-                        z.extract(artifact, os.path.join(dirname, artifact))
-            return True
-        return False
-
+                return zip_fingerprint_file
+        return None
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply Cython/Build/Cache.py]: done.

File: Cython/Build/Cache.py [hunk 4]
```diff
@@ -131,44 +144,58 @@ def lookup_cache(self, c_file, fingerprint):
...
-            for attr in ('c_file', 'h_file', 'api_file', 'i_file')]))
+        artifacts = list(
+            filter(
+                None,
+                [
+                    getattr(compilation_result, attr, None)
+                    for attr in ("c_file", "h_file", "api_file", "i_file")
+                ],
+            )
+        )
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply Cython/Build/Cache.py]: fixed.

File: Cython/Build/Cache.py [hunk 5]
```diff
@@ -94,20 +110,18 @@ def transitive_fingerprint(self, filename, dependencies, compilation_options, la
...
             # in the fingerprint. We do not iterate over module.__dict__ and
             # include almost everything here as users might extend Extension
             # with arbitrary (random) attributes that would lead to cache
             # misses.
-            m.update(
-                str((language, py_limited_api,np_pythran)).encode('UTF-8')
-            )
-            m.update(compilation_options.get_fingerprint().encode('UTF-8'))
+            m.update(flags.get_fingerprint())
+            m.update(compilation_options.get_fingerprint().encode("UTF-8"))
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply Cython/Build/Cache.py]: done.

**Comment by scoder** [comment]
Nice, that looks much better.

**Pull request merged**